kiwi_lint/function_based_views.py

```diff
@@ -0,0 +1,133 @@
+
+    def _get_django_project_apps(self):
+        installed_apps = deepcopy(settings.INSTALLED_APPS)  # do not mutate the original
+
+        for filter_func in self.installed_apps_function_filters:
+            # allow void filters that just mutate the list
+            res = list(filter_func(installed_apps) or [])
+            installed_apps = res if res else installed_apps
+
+        return installed_apps
+
+    @staticmethod
+    def _resolve_app_configs(installed_apps):
+        import_paths_by_configs = {
+            f'{config.__class__.__module__}.{config.__class__.__name__}': config.name
+            for config in django_apps.get_app_configs()
+        }
+
+        resolved = []
+        for name in installed_apps:
+            resolved.append(import_paths_by_configs.get(name, name))  # get(name, default=name)
+
+        return resolved
+
+    @classmethod
+    def _get_url_view_mapping(cls, urlpatterns):
+        def helper(urlpatterns, prefix='^', acc=None):
+            if acc is None:
+                acc = {}
+
+            for url in urlpatterns:
+                if isinstance(url, URLPattern):
+                    key = prefix + url.pattern.regex.pattern.strip('^')
+                    acc[key] = (url.callback.__module__, url.callback.__name__)
+
+                elif isinstance(url, URLResolver):
+                    helper(url.url_patterns, prefix + url.pattern.regex.pattern.strip('^$'), acc)
+            return acc
+
+        return helper(urlpatterns)
+
+    @staticmethod
+    def _prune_url_mapping(url_mapping, installed_apps):
+        pruned = url_mapping.__class__()
+        for k, (module, name) in url_mapping.items():
+            if any(module.startswith(app) for app in installed_apps):
+                pruned[k] = (module, name)
+
+        return pruned
+
+    def visit_module(self, module):
+        if module.name in self.view_files:
+            self.visit_views_module(module)
+
+    def visit_views_module(self, module):
+        """Called when entering a module with a registered view inside it."""
+
+
+class FunctionBasedViewChecker(DjangoViewsVisiter):
+    __implements__ = (interfaces.IAstroidChecker, )
+
+    name = 'django-function-based-view-checker'
+
+    msgs = {'R4611': ('Function based view used, use class based',
+                      'non-function-based-view-required',
+                      'Function based views are the path to the Dark side!')}
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.views_by_module = self.group_views_by_module(self.url_mapping.values())
+
+    @staticmethod
+    def group_views_by_module(module_view_name_pairs):
+        by_module = {}
+        for module, identifier in module_view_name_pairs:
+            by_module.setdefault(module, set()).add(identifier)
+
+        return by_module
+
+    def visit_views_module(self, module):
+        for func_def in module.nodes_of_class(astroid.FunctionDef):
```

Comment:
You can simplify this by just having `visit_module` to make sure you are in a module which has views (will set a boolean flag) and `visit_functiondef` to examine if this is a FBV and call `add_message`.